### PR TITLE
[GLUTEN-2360][CORE] feat: adding more fallback policies

### PR DESCRIPTION
**Pull request by zhouyuan**
## What changes were proposed in this pull request?

This patch adds more fallback policies to run the query if there's one fallback:

```
operator: fallback per operator if not supported
stage: fallback per stage if number of unsupported operators > threashold
query: fallback whole query if there are more than one operators not supported

```
Fixes: #2360 

Signed-off-by: Yuan Zhou <yuan.zhou@intel.com>
## How was this patch tested?

UT

### CODE REVIEW

**Comment by baibaichen** [comment]
@zhouyuan 
 Why non-AQE？ I think it is better to apply to both.

**Comment by zhouyuan** [comment]
> @zhouyuan Why non-AQE？ I think it is better to apply to both.

yes, let me try to combine the code on AQE and non-AQE code path

**Comment by zuochunwei** [comment]
good

File: gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala [hunk 1]
```diff
@@ -752,6 +756,27 @@ case class ColumnarOverrideRules(session: SparkSession)
...
+          fallbacks = fallbacks + 1
+        // Possible fallback for leaf node.
+        case leafPlan: LeafExecNode if !leafPlan.isInstanceOf[GlutenPlan] =>
+          fallbacks = fallbacks + 1
+        case _ =>
+      }
+      plan.children.map(p => countFallback(p))
+    }
+    countFallback(plan)
+    fallbacks > 1
```
Comment (zzcclp) [inline gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala:777]: can we consider to use `wholeStageFallbackThreshold` as threshold too for query policy ?

File: shims/common/src/main/scala/io/glutenproject/GlutenConfig.scala [hunk 2]
```diff
@@ -779,6 +781,16 @@ object GlutenConfig {
       .longConf
       .createWithDefault(100 * 1000 * 1000)
 
+  val COLUMNAR_FALLBACK_POLICY =
+    buildConf("spark.gluten.sql.columnar.fallback.policy")
+      .internal()
+      .doc("The fallback policy in gluten. By default, 'operator' means fallback each " +
+        "operator, 'stage' means fallback each stage, 'query' means fallback query")
```
Comment (zzcclp) [inline shims/common/src/main/scala/io/glutenproject/GlutenConfig.scala:788]: stage fallback only for AQE on, right ?

Comment (zhouyuan) [reply shims/common/src/main/scala/io/glutenproject/GlutenConfig.scala:788]: right, the current impl replies AQE to check if it's fully supported

File: gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala (hunk 1 above)
Comment (zhouyuan) [reply gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala:777]: yes, let me try to add another configurable value for this

File: gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala [hunk 3]
```diff
@@ -752,6 +758,30 @@ case class ColumnarOverrideRules(session: SparkSession)
...
+          fallbacks = fallbacks + 1
+        // Possible fallback for leaf node.
+        case leafPlan: LeafExecNode if !leafPlan.isInstanceOf[GlutenPlan] =>
+          fallbacks = fallbacks + 1
+        case _ =>
+      }
+      plan.children.map(p => countFallback(p))
+    }
+    countFallback(plan)
+    fallbacks > queryFallbackThreshold
```
Comment (zzcclp) [inline gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala:782]: use `>=`

File: gluten-ut/spark32/src/test/scala/org/apache/spark/sql/execution/FallbackStrategiesSuite.scala [hunk 4]
```diff
@@ -25,8 +25,25 @@ import org.apache.spark.sql.GlutenSQLTestsTrait
 
 class FallbackStrategiesSuite extends GlutenSQLTestsTrait {
 
+  test("Fall back the whole query if one unsupported") {
```
Comment (zzcclp) [inline gluten-ut/spark32/src/test/scala/org/apache/spark/sql/execution/FallbackStrategiesSuite.scala:28]: add a ut for query fallback threshold

File: gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala [hunk 5]
```diff
@@ -793,7 +823,10 @@ case class ColumnarOverrideRules(session: SparkSession)
 
   override def postColumnarTransitions: Rule[SparkPlan] = plan => PhysicalPlanSelector.
     maybe(session, plan) {
-      if (isAdaptiveContext && fallbackWholeStage(plan)) {
+      if (fallbackPolicy == "query"&& !isAdaptiveContext  && fallbackWholeQuery(plan)) {
```
Comment (zzcclp) [inline gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala:826]: add a blank between `fallbackPolicy == "query"` and `&&`

File: gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala [hunk 6]
```diff
@@ -752,6 +758,30 @@ case class ColumnarOverrideRules(session: SparkSession)
     fallbacks >= wholeStageFallbackThreshold
   }
 
+  def fallbackWholeQuery(plan: SparkPlan): Boolean = {
```
Comment (zzcclp) [inline gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala:761]: the logical of this function is the same as fallbackWholeStage, can it be removed ?

Comment (zhouyuan) [reply gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala:761]: indeed, will merged in to one function

**Review by zzcclp — APPROVED** [review]
LGTM

**Comment by baibaichen** [comment]
@zhouyuan I don't understand the codes before trying it, please merge it.

**Pull request merged**